app/src/main/java/com/github/swent/echo/data/repository/RepositoryImpl.kt

```diff
@@ -288,20 +408,32 @@ class RepositoryImpl(
         if (isOffline()) {
             throw RepositoryStoreWhileNoInternetException("UserProfilePicture")
         }
-        remoteDataSource.setUserProfilePicture(userId, picture)
+        try {
+            remoteDataSource.setUserProfilePicture(userId, picture)
+        } catch (e: RemoteDataSourceRequestMaxRetryExceededException) {
+            throw RepositoryStoreWhileNoInternetException("UserProfilePicture")
+        }
         fileCache.set("$userId.jpeg", picture)
     }
 
     override suspend fun deleteUserProfilePicture(userId: String) {
         if (isOffline()) {
             throw RepositoryStoreWhileNoInternetException("UserProfilePicture")
         }
-        remoteDataSource.deleteUserProfilePicture(userId)
+        try {
+            remoteDataSource.deleteUserProfilePicture(userId)
+        } catch (e: RemoteDataSourceRequestMaxRetryExceededException) {
+            throw RepositoryStoreWhileNoInternetException("UserProfilePicture")
+        }
         fileCache.delete("$userId.jpeg")
     }
 }
 
-class RepositoryStoreWhileNoInternetException(objectTryingToStore: String) :
-    Exception(
-        "Storing/Deleting a " + objectTryingToStore + " while the App is offline is not supported"
-    )
+private fun <DataObject> DataObject.getId(): String =
+    when (this) {
+        is Association -> this.associationId
+        is Event -> this.eventId
+        is Tag -> this.tagId
+        is UserProfile -> this.userId
+        else -> ""
+    }
```

Comment:
Instead of using a generic type `DataObject`, I'd suggest you use a sealed class `DataModel` for all our data model, this would make it typesafe and avoid the `else` branch.

Reply:
I tried creating a supertype for the dataclasses, but somehow dataclasses can't inherit in Kotlin. I was not aware that I could use sealed classes for this. Indeed it is the best solution for this! Thank you very much!